File: src/deploy/yolov5_face/python/yolov5_face_detector.py
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
#
# SPDX-License-Identifier: Apache-2.0

"""
YOLOv5-Face Detector Implementation
Aligned with original code from k3/ai_web/model_zoo/examples/CV/yolov5-face/python/utils.py
"""

import cv2
import numpy as np
import onnxruntime as ort
import spacemit_ort  # noqa: F401 - register SpaceMITExecutionProvider
import time
from typing import List, Tuple, Optional, Dict, Any
from pathlib import Path

from core import BaseModel
from common import (
    sigmoid, letterbox, xywh2xyxy, nms, scale_coords
)
from core.python.vision_model_exceptions import (
    ModelNotFoundError,
    ModelLoadError,
    ModelInferenceError,
)


class YOLOv5FaceDetector(BaseModel):
    """
    YOLOv5-Face detector for face detection.
    Same interface as original Yolov5FaceDetection class.
    """

    def __init__(self, model_path: str,
                 conf_thres: float = 0.25,
                 iou_thres: float = 0.45,
                 num_threads: int = 4,
                 **kwargs):
        """
        Initialize YOLOv5-Face detector.
        Same as original code.

        Args:
            model_path: Path to ONNX model file
            conf_thres: Confidence threshold for filtering detections
            iou_thres: IoU threshold for NMS
            num_threads: Number of threads for ONNX Runtime
            **kwargs: Additional parameters
        """
        self.conf_thres = conf_thres
        self.iou_thres = iou_thres
        self.num_threads = num_threads
        super().__init__(model_path, **kwargs)

    def _load_model(self, **kwargs):
        """Load YOLOv5-Face ONNX model."""
        # Check if model file exists
        model_path_obj = Path(self.model_path)
        if not model_path_obj.exists():
            raise ModelNotFoundError(
                model_path=self.model_path,
                message=f"YOLOv5-Face 模型文件不存在: {self.model_path}"
            )

        # Create session options
        session_options = ort.SessionOptions()
        session_options.intra_op_num_threads = self.num_threads

        # Load model
        try:
            providers = kwargs.get('providers', ['SpaceMITExecutionProvider'])
            self.session = ort.InferenceSession(
                self.model_path,
                sess_options=session_options,
                providers=providers
            )
        except Exception as e:
            raise ModelLoadError(
                model_path=self.model_path,
                reason=str(e),
                message=f"YOLOv5-Face 模型加载失败: {self.model_path} - {e}"
            )

        # Get input/output info - same as original: shape[2:4] not tuple
        try:
            self.input_name = self.session.get_inputs()[0].name
            self.input_shape = self.session.get_inputs()[0].shape[2:4]
        except Exception as e:
            raise ModelLoadError(
                model_path=self.model_path,
                reason=f"无法获取模型输入/输出信息: {e}",
                message=f"YOLOv5-Face 模型初始化失败: {self.model_path} - {e}"
            )


    def preprocess(self, image: np.ndarray) -> np.ndarray:
        """
        Preprocess image for YOLOv5-Face inference.
        Same as original code: process_img method.

        Args:
            img: Input image in BGR format

        Returns:
            Preprocessed tensor ready for inference
        """
        img = letterbox(image, self.input_shape)[0]
        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        img = img.transpose(2, 0, 1).copy()

        # Same as original code: cv2.normalize instead of direct division
        img = cv2.normalize(img, None, 0, 1, cv2.NORM_MINMAX, dtype=cv2.CV_32F)
        img = img[None]
        return img


    def postprocess(self, outputs: List[np.ndarray], **kwargs) -> np.ndarray:
        """
        Postprocess YOLOv5-Face outputs.
        Same as original code: post_process method.

        Args:
            output: Raw model outputs

        Returns:
            Processed predictions
        """
        init_anchors = np.array([
            4., 5., 8., 10., 13., 16., 23., 29.,
            43., 55., 73., 105., 146., 217.,
            231., 300., 335., 433.
        ]).astype(np.float32).reshape(3, 1, 3, 1, 1, 2)

        strides = [8, 16, 32]
        pred_list = []

        for idx, pred in enumerate(outputs):
            y = np.full_like(pred[..., :6], 0)
            bs, _, ny, nx, _ = pred.shape
            grid = self._make_grid(nx, ny)

            y_tmp = sigmoid(np.concatenate([pred[..., :5], pred[..., 15:]], axis=-1))
            y[..., :2] = (y_tmp[..., :2] * 2. - 0.5 + grid) * strides[idx]
            y[..., 2:4] = (y_tmp[..., 2:4] * 2) ** 2 * init_anchors[idx]
            y[..., 4:] = y_tmp[..., 4:]
            pred_list.append(y.reshape(bs, -1, 6))

        pred_result = np.concatenate(pred_list, 1)
        return pred_result

    def infer(self, img_src: np.ndarray) -> Tuple[List[Dict[str, Any]], List[np.ndarray]]:
        """
        Run YOLOv5-Face inference on an image.
        Same as original code: returns detections and face_images.

        Args:
            img_src: Input image in BGR format

        Returns:
            Tuple of (detections, face_images):
            - detections: List of dicts with "bbox" and "confidence"
            - face_images: List of cropped face images
        """
        img = img_src.copy()

        # Preprocess
        try:
            img = self.preprocess(img)
        except Exception as e:
            raise ModelInferenceError(
                model_name="YOLOv5-Face",
                reason=f"图像预处理失败: {e}",
                message=f"YOLOv5-Face 预处理错误: {e}"
            )

        # Run inference
        try:
            model_infer = self.session.run(None, {self.input_name: img})
        except Exception as e:
            raise ModelInferenceError(
                model_name="YOLOv5-Face",
                reason=f"模型推理失败: {e}",
                message=f"YOLOv5-Face 推理错误: {e}"
            )

        # Postprocess
        try:
            pred = self.postprocess(model_infer)
            pred = self.non_max_suppression_face(pred, self.conf_thres, self.iou_thres)
        except Exception as e:
            raise ModelInferenceError(
                model_name="YOLOv5-Face",
                reason=f"结果后处理失败: {e}",
                message=f"YOLOv5-Face 后处理错误: {e}"
            )

        detections = []
        face_images = []
        if pred is not None and len(pred) > 0 and len(pred[0]) > 0:
            for i in range(len(pred[0])):
                # Same as original: use img_src.shape (not shape[:2])
                bbox = scale_coords(self.input_shape, pred[0][i][:4], img_src.shape).round()
                conf = float(pred[0][i][4]) if len(pred[0][i]) > 4 else 1.0
                face_images.append(img_src[int(bbox[1]):int(bbox[3]), int(bbox[0]):int(bbox[2])])
                detections.append({
                    "bbox": bbox.tolist(),
                    "confidence": conf
                })

        return detections, face_images

    def _make_grid(self, nx: int, ny: int) -> np.ndarray:
        """Create grid for anchor processing. Same as original _make_grid function."""
        xv, yv = np.meshgrid(np.arange(ny), np.arange(nx))
        return np.stack((xv, yv), 2).reshape((1, 1, ny, nx, 2)).astype(np.float32)

    def non_max_suppression_face(self, prediction: np.ndarray,
                                  conf_thres: float = 0.25,
                                  iou_thres: float = 0.45,
                                  classes: Optional[List[int]] = None,
                                  agnostic: bool = False,
                                  labels: Tuple[str, ...] = ()) -> List[np.ndarray]:
        """
        Apply NMS for face detection.
        Same as original code: non_max_suppression_face static method.
        """
        nc = prediction.shape[2] - 5  # number of classes
        xc = prediction[..., 4] > conf_thres  # candidates

        # Settings
        max_wh = 4096  # (pixels) maximum box width and height
        time_limit = 10.0  # seconds to quit after

        t = time.time()
        output = [np.zeros((0, 6))] * prediction.shape[0]
        for xi, x in enumerate(prediction):  # image index, image inference
            x = x[xc[xi]]  # confidence
            # Cat apriori labels if autolabelling
            if labels and len(labels[xi]):
                lbl = labels[xi]
                v = np.zeros((len(lbl), nc + 5))
                v[:, :4] = lbl[:, 1:5]  # box
                v[:, 4] = 1.0  # conf
                v[range(len(lbl)), lbl[:, 0].astype(int) + 5] = 1.0  # cls
                x = np.concatenate((x, v), 0)
            # If none remain process next image
            if not x.shape[0]:
                continue

            # Compute conf
            x[:, 5:] *= x[:, 4:5]  # conf = obj_conf * cls_conf

            box = xywh2xyxy(x[:, :4])
            conf = x[:, 5:].max(1).reshape(-1, 1)
            class_idx = np.argmax(x[:, 5:], axis=1).reshape(-1, 1).astype(np.float32)
            x = np.concatenate((box, conf, class_idx), 1)[conf.reshape(-1) > conf_thres]

            # If none remain process next image
            n = x.shape[0]  # number of boxes
            if not n:
                continue

            # Batched NMS
            c = x[:, 5:6] * (0 if agnostic else max_wh)  # classes
            boxes, scores = x[:, :4] + c, x[:, 4]  # boxes (offset by class), scores
            i = nms(boxes, scores, iou_thres)  # NMS

            output[xi] = x[i]
            if (time.time() - t) > time_limit:
                break  # time limit exceeded

        return output

File: src/deploy/yolov5_face/cpp/yolov5_face_detector.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include "yolov5_face_detector.h"

#include <chrono>
#include <algorithm>
#include <cmath>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

#include "common.h"
#include "vision_model_config.h"
#include "vision_model_factory.h"

namespace vision_deploy {

std::unique_ptr<vision_core::BaseModel> YOLOv5FaceDetector::create(const YAML::Node& config, bool lazy_load) {
    std::string model_path = vision_core::yaml_utils::getString(config, "model_path");
    if (model_path.empty()) {
        throw std::runtime_error("model_path not found in config for YOLOv5FaceDetector");
    }

    YAML::Node default_params = config["default_params"];
    if (!default_params) {
        throw std::runtime_error("default_params not found in config for YOLOv5FaceDetector");
    }

    float conf_threshold = vision_core::yaml_utils::getFloat(default_params, "conf_threshold", 0.25f);
    float iou_threshold = vision_core::yaml_utils::getFloat(default_params, "iou_threshold", 0.45f);
    int num_threads = vision_core::yaml_utils::getInt(default_params, "num_threads", 4);
    std::string provider = vision_core::yaml_utils::getProvider(config);

    return std::make_unique<YOLOv5FaceDetector>(
        model_path, conf_threshold, iou_threshold, num_threads, lazy_load, provider);
}

YOLOv5FaceDetector::YOLOv5FaceDetector(const std::string& model_path,
                                        float conf_threshold,
                                        float iou_threshold,
                                        int num_threads,
                                        bool lazy_load,
                                        const std::string& provider)
    : BaseModel(model_path, lazy_load),
        conf_threshold_(conf_threshold),
        iou_threshold_(iou_threshold),
        num_threads_(num_threads),
        provider_(provider) {
    if (!lazy_load) {
        load_model();
    }
}

void YOLOv5FaceDetector::load_model() {
    init_session(num_threads_, provider_);
    model_loaded_ = true;
}

cv::Mat YOLOv5FaceDetector::preprocess(const cv::Mat& image) {
    if (image.empty()) {
        throw std::runtime_error("Input image is empty");
    }

    ensure_model_loaded();

    int inputWidth = static_cast<int>(input_shape_[3]);  // width
    int inputHeight = static_cast<int>(input_shape_[2]);  // height

    // Use common letterbox function (already converts BGR to RGB)
    cv::Mat padded = vision_common::letterbox(image,
                                            std::make_pair(inputHeight, inputWidth));
    return cv::dnn::blobFromImage(padded, 1.0/255.0,
        cv::Size(inputWidth, inputHeight),
        cv::Scalar(0, 0, 0), true, false, CV_32F);
}

std::vector<vision_common::Result> YOLOv5FaceDetector::detect(const cv::Mat& image) {
    ensure_model_loaded();
    reset_runtime_profile();
    const auto t0 = std::chrono::steady_clock::now();

    cv::Size orig_size = image.size();

    // Preprocess
    const auto t_pre0 = std::chrono::steady_clock::now();
    cv::Mat inputTensor = preprocess(image);
    const auto t_pre1 = std::chrono::steady_clock::now();
    set_runtime_preprocess_ms(std::chrono::duration<double, std::milli>(t_pre1 - t_pre0).count());

    // Run inference using base class method
    const auto t_infer0 = std::chrono::steady_clock::now();
    std::vector<Ort::Value> outputs = run_session(inputTensor);
    const auto t_infer1 = std::chrono::steady_clock::now();
    set_runtime_model_infer_ms(std::chrono::duration<double, std::milli>(t_infer1 - t_infer0).count());

    // Postprocess
    const auto t_post0 = std::chrono::steady_clock::now();
    std::vector<vision_common::Result> results = postprocess(outputs, orig_size);
    const auto t_post1 = std::chrono::steady_clock::now();
    set_runtime_postprocess_ms(std::chrono::duration<double, std::milli>(t_post1 - t_post0).count());

    const auto t1 = std::chrono::steady_clock::now();
    set_runtime_total_ms(std::chrono::duration<double, std::milli>(t1 - t0).count());

    return results;
}

std::vector<vision_core::ModelCapability> YOLOv5FaceDetector::get_capabilities() const {
    return {
        vision_core::ModelCapability::kImageInput,
        vision_core::ModelCapability::kDraw};
}

std::vector<vision_common::Result> YOLOv5FaceDetector::postprocess(
    std::vector<Ort::Value>& outputs,
    const cv::Size& orig_size) {

    // Initialize anchors (same as Python code: reshape(3, 1, 3, 1, 1, 2))
    // Reshape to (3, 1, 3, 1, 1, 2) means: 3 scales, each with 3 anchors, each anchor has (w, h)
    // Simplified to (3, 3, 2) for easier access: [scale][anchor][w_or_h]
    const float init_anchors[3][3][2] = {
        {{4.0f, 5.0f}, {8.0f, 10.0f}, {13.0f, 16.0f}},
        {{23.0f, 29.0f}, {43.0f, 55.0f}, {73.0f, 105.0f}},
        {{146.0f, 217.0f}, {231.0f, 300.0f}, {335.0f, 433.0f}}
    };

    const int strides[3] = {8, 16, 32};

    // Optimized: Pre-compute logit threshold to avoid unnecessary sigmoid calculations
    // Reverse sigmoid: raw_conf_threshold = -log(1.0/conf_threshold - 1)
    // If raw value < threshold, sigmoid(raw) < conf_threshold, so we can skip
    float raw_conf_threshold = -std::log(1.0f / conf_threshold_ - 1.0f);

    // Optimized: Pre-allocate memory and use fixed-size struct instead of vector
    // Estimate total predictions: 3 scales * channels * ny * nx
    // Reserve space to avoid reallocations
    std::vector<Prediction> pred_list;
    size_t estimated_size = 0;

    // First pass: calculate total size for pre-allocation
    for (size_t idx = 0; idx < outputs.size() && idx < 3; ++idx) {
        auto tensor_info = outputs[idx].GetTensorTypeAndShapeInfo();
        std::vector<int64_t> dims = tensor_info.GetShape();
        int channels = static_cast<int>(dims[1]);  // 3 anchors
        int ny = static_cast<int>(dims[2]);
        int nx = static_cast<int>(dims[3]);
        estimated_size += channels * ny * nx;
    }
    pred_list.reserve(estimated_size);

    // Process each output (3 scales)
    for (size_t idx = 0; idx < outputs.size() && idx < 3; ++idx) {
        const float* pred_data = outputs[idx].GetTensorMutableData<float>();
        auto tensor_info = outputs[idx].GetTensorTypeAndShapeInfo();
        std::vector<int64_t> dims = tensor_info.GetShape();

        int bs = static_cast<int>(dims[0]);
        int channels = static_cast<int>(dims[1]);  // 3 anchors
        int ny = static_cast<int>(dims[2]);
        int nx = static_cast<int>(dims[3]);
        int features = static_cast<int>(dims[4]);  // 16 features

        // Create grid
        std::vector<float> grid = make_grid(nx, ny);

        // Process predictions for batch 0
        for (int anchor = 0; anchor < channels; ++anchor) {
            for (int h = 0; h < ny; ++h) {
                for (int w = 0; w < nx; ++w) {
                    size_t base_idx = anchor * ny * nx * features +
                                        h * nx * features + w * features;

                    // Optimized: Early exit if obj_conf (index 4) is below threshold
                    // Compare raw logit value first to avoid expensive sigmoid calculation
                    float raw_obj_conf = pred_data[base_idx + 4];
                    if (raw_obj_conf < raw_conf_threshold) {
                        continue;  // Skip this prediction, no need to compute sigmoid
                    }

                    // Only compute sigmoid for predictions that pass the threshold check
                    // Same as Python: y_tmp = sigmoid(np.concatenate([pred[..., :5], pred[..., 15:]], axis=-1))
                    float x_tmp = vision_common::sigmoid(pred_data[base_idx]);
                    float y_tmp = vision_common::sigmoid(pred_data[base_idx + 1]);
                    float w_tmp = vision_common::sigmoid(pred_data[base_idx + 2]);
                    float h_tmp = vision_common::sigmoid(pred_data[base_idx + 3]);
                    float obj_conf = vision_common::sigmoid(raw_obj_conf);
                    float cls_conf = vision_common::sigmoid(pred_data[base_idx + 15]);  // Last feature

                    // Decode coordinates: (y_tmp * 2. - 0.5 + grid) * strides[idx]
                    // Python: y[..., :2] = (y_tmp[..., :2] * 2. - 0.5 + grid) * strides[idx]
                    // grid has shape (1, 1, ny, nx, 2), accessed as grid[h][w] = [x, y]
                    int grid_idx = (h * nx + w) * 2;
                    float x = (x_tmp * 2.0f - 0.5f + grid[grid_idx]) * strides[idx];
                    float y = (y_tmp * 2.0f - 0.5f + grid[grid_idx + 1]) * strides[idx];

                    // Decode width/height: (y_tmp[..., 2:4] * 2) ** 2 * init_anchors[idx]
                    // Python: y[..., 2:4] = (y_tmp[..., 2:4] * 2) ** 2 * init_anchors[idx]
                    // init_anchors[idx] has shape (1, 3, 1, 1, 2), simplified to (3, 2) for easier access
                    float w_box = std::pow(w_tmp * 2.0f, 2) * init_anchors[idx][anchor][0];
                    float h_box = std::pow(h_tmp * 2.0f, 2) * init_anchors[idx][anchor][1];

                    // Store: [x, y, w, h, obj_conf, cls_conf] in xywh format (same as Python)
                    // Python postprocess returns (bs, -1, 6) with [x, y, w, h, obj_conf, cls_conf]
                    // Conversion to xyxy happens in NMS
                    // Optimized: Use fixed-size struct instead of vector to avoid heap allocation
                    pred_list.emplace_back(x, y, w_box, h_box, obj_conf, cls_conf);
                }
            }
        }
    }

    // Apply NMS
    std::vector<vision_common::Result> results = non_max_suppression_face(pred_list);

    // Scale coordinates back to original image size
    // Python: scale_coords(self.input_shape, pred[0][i][:4], img_src.shape)
    // self.input_shape is shape[2:4] = [height, width] from Python
    // img_src.shape is (height, width, channels)
    // In C++: input_shape_ is [batch, channels, height, width] from ONNX
    // Python uses shape[2:4] which is indices 2 and 3 (height, width)
    // Same as yolov8_detector.cpp: input_shape_[2] = height, input_shape_[3] = width
    int inputHeight = static_cast<int>(input_shape_[2]);  // height (index 2)
    int inputWidth = static_cast<int>(input_shape_[3]);   // width (index 3)
    cv::Size input_shape(inputWidth, inputHeight);  // cv::Size is (width, height)

        for (auto& result : results) {
        float coords[4] = {result.x1, result.y1, result.x2, result.y2};
        vision_common::scale_coords(input_shape, coords, orig_size);
        result.x1 = coords[0];
        result.y1 = coords[1];
        result.x2 = coords[2];
        result.y2 = coords[3];
    }

    return results;
}

std::vector<vision_common::Result> YOLOv5FaceDetector::non_max_suppression_face(
    const std::vector<Prediction>& predictions) {

    if (predictions.empty()) {
        return std::vector<vision_common::Result>();
    }

    // First filter: obj_conf > conf_threshold (same as Python: xc = prediction[..., 4] > conf_thres)
    // Optimized: Pre-allocate filtered_preds to avoid reallocations
    std::vector<Prediction> filtered_preds;
    filtered_preds.reserve(predictions.size());
    for (const auto& pred : predictions) {
        if (pred.obj_conf > conf_threshold_) {  // obj_conf
            filtered_preds.push_back(pred);
        }
    }

    if (filtered_preds.empty()) {
        return std::vector<vision_common::Result>();
    }

    // Compute final confidence: conf = obj_conf * cls_conf (same as Python: x[:, 5:] *= x[:, 4:5])
    // Convert xywh to xyxy (same as Python: box = xywh2xyxy(x[:, :4]))
    // Optimized: Pre-allocate candidates to avoid reallocations
    std::vector<vision_common::Result> candidates;
    candidates.reserve(filtered_preds.size());
    for (const auto& pred : filtered_preds) {
        float final_conf = pred.obj_conf * pred.cls_conf;  // Final confidence

        // Second filter: final_conf > conf_threshold (same as Python: [conf.reshape(-1) > conf_thres])
        if (final_conf > conf_threshold_) {
            // Convert xywh to xyxy (same as Python: box = xywh2xyxy(x[:, :4]))
            float xywh[4] = {pred.x, pred.y, pred.w, pred.h};
            float xyxy[4];
            vision_common::xywh2xyxy(xywh, xyxy);

            vision_common::Result result;
            result.x1 = xyxy[0];
            result.y1 = xyxy[1];
            result.x2 = xyxy[2];
            result.y2 = xyxy[3];
            result.score = final_conf;  // Use final confidence
            result.label = -1;  // Face class (no label display)
            candidates.push_back(result);
        }
    }

    // Apply NMS using common function
    if (candidates.empty()) {
        return std::vector<vision_common::Result>();
    }

    std::vector<cv::Rect2f> boxes;
    std::vector<float> scores;
    boxes.reserve(candidates.size());
    scores.reserve(candidates.size());

    for (const auto& cand : candidates) {
        boxes.push_back(cv::Rect2f(cand.x1, cand.y1, cand.x2 - cand.x1, cand.y2 - cand.y1));
        scores.push_back(cand.score);
    }

    std::vector<int> keep_indices = vision_common::nms(boxes, scores, iou_threshold_);

    std::vector<vision_common::Result> final_results;
    final_results.reserve(keep_indices.size());
    for (int idx : keep_indices) {
        final_results.push_back(candidates[idx]);
    }

    return final_results;
}

std::vector<float> YOLOv5FaceDetector::make_grid(int nx, int ny) {
    // Flat array: [w0, h0, w1, h1, ...] — 2 floats per grid cell
    std::vector<float> grid(static_cast<size_t>(ny * nx * 2));
    for (int h = 0; h < ny; ++h) {
        for (int w = 0; w < nx; ++w) {
            size_t idx = static_cast<size_t>((h * nx + w) * 2);
            grid[idx] = static_cast<float>(w);
            grid[idx + 1] = static_cast<float>(h);
        }
    }
    return grid;
}


// Self-registration (runs at program startup)
static vision_core::ModelRegistrar<YOLOv5FaceDetector> registrar("YOLOv5FaceDetector");

}  // namespace vision_deploy

